File: solo/puzzles/easy/max-area/max_area.rs
use std::io;

macro_rules! parse_input {
   ($x:expr, $t:ident) => ($x.trim().parse::<$t>().unwrap())
}

fn main() {
   let mut input_line = String::new();
   io::stdin().read_line(&mut input_line).unwrap();
   let n = parse_input!(input_line, i32);
   let mut inputs = String::new();
   io::stdin().read_line(&mut inputs).unwrap();
   let mut segments = Vec::new();
   for i in inputs.split_whitespace() {
      let a = parse_input!(i, i32);
      segments.push(a);
   }

   let mut left = 0;
   let mut right = segments.len() - 1;
   let mut max_area = 0;
   while left != right {
      let height = i32::min(segments[left], segments[right]);
      let width = (right - left) as i32;
      let area = height * width;
      if area > max_area {
         max_area = area;
      }
      if segments[left] < segments[right] {
         left += 1;
      } else {
         right -= 1;
      }
   }
   println!("{}", max_area);
}

File: solo/puzzles/easy/max-area/max_area.java
import java.util.*;

class Solution {

    public static void main(String args[]) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        List<Integer> arr = new ArrayList<Integer>();
        for (int i = 0; i < n; i++) {
            int a = in.nextInt();
            arr.add(a);
        }

        int left = 0;
        int right = arr.size() - 1;

        int maxArea = 0;
        while (left != right)
        {
            int area = Math.min(arr.get(left), arr.get(right)) * (right - left);
            if (area > maxArea)
            {
                maxArea = area;
            }

            if (arr.get(left) < arr.get(right))
            {
                ++left;
            }
            else
            {
                --right;
            }
        }

        System.out.println(maxArea);
    }
}
